src/main/java/io/openliberty/tools/intellij/lsp4jakarta/lsp4ij/PropertiesManagerForJakarta.java

```diff
@@ -118,6 +124,73 @@ private void collectDiagnostics(String uri, IPsiUtils utils, DocumentFormat docu
+     * @param uri             - String representing file from which to derive project
+     *                        classpath
+     * @param snippetContexts - get all the context fields from the snippets and
+     *                        check if they exist in this method
+     * @param project         - the IntelliJ project containing the uri
+     * @return List<String>
+     */
+    public List<String> getExistingContextsFromClassPath(String uri, List<String> snippetContexts, Project project) {
+        // ask the Java component if the classpath of the current module contains the specified Jakarta types.
+        JavaPsiFacade javaFacade = JavaPsiFacade.getInstance(project);
+        GlobalSearchScope scope = GlobalSearchScope.allScope(project);
+        List<String> validCtx = new ArrayList<String>();
+        if (javaFacade != null && scope != null) {
+            for (String typeCtx : snippetContexts) {
+                Object type = ApplicationManager.getApplication().runReadAction((Computable<Object>) () -> javaFacade.findClass(typeCtx, scope));
+                validCtx.add(type != null ? typeCtx : null); // list will be the same size as input
+            }
+        } else {
+            // Error: none of these contexts will add to the completions
+            for (String typeCtx : snippetContexts) {
+                validCtx.add(null); // list will be the same size as input
+            }
+        }
+
+        // FOR NOW, append package name and class name to the list in order for LS to
+        // resolve ${packagename} and ${classname} variables
+        PsiFile typeRoot = ApplicationManager.getApplication().runReadAction((Computable<PsiFile>) () -> resolveTypeRoot(uri, project));
+        DumbService.getInstance(project).runReadActionInSmartMode(() -> {
+            String className = "className";
+            String packageName = "packageName";
+            if (typeRoot instanceof PsiJavaFile) {
+                PsiJavaFile javaFile = (PsiJavaFile)typeRoot;
+                PsiClass[] classes = javaFile.getClasses();
+                if (classes.length > 0) {
+                    className = classes[0].getName();
+                } else {
+                    className = javaFile.getName();
+                    if (className.endsWith(".java") == true) {
+                        className = className.substring(0, className.length() - 5);
+                    }
+                }
+                packageName = javaFile.getPackageName();
+                if (packageName == null || packageName.isEmpty()) {
+                    String path = javaFile.getParent().getVirtualFile().getCanonicalPath(); // f=/U/me/proj/src/main/java/pkg  not /cls.java
+                    VirtualFile[] contentRoots = ProjectRootManager.getInstance(project).getContentSourceRoots();
+                    for (VirtualFile vf : contentRoots) {
+                        String vfp = vf.getCanonicalPath();
+                        if (path.startsWith(vfp)) {
```

Comment:
In https://github.com/OpenLiberty/liberty-tools-intellij/pull/183#issuecomment-1333029972, I found that converting from IntelliJ VritualFile to paths worked well with VirtualFile.toNioPath(). Not sure if that would simplify things here, just to consider.